chainerx_cc/examples/mnist_cc/train_mnist.cc

```diff
@@ -0,0 +1,203 @@
+            params_.emplace_back(chainerx::Zeros({n_out}, chainerx::TypeToDtype<float>));
+        }
+
+        for (const chainerx::Array& param : params_) {
+            param.RequireGrad();
+        }
+    }
+
+private:
+    int64_t n_in_;
+    int64_t n_hidden_;
+    int64_t n_out_;
+    int64_t n_layers_;
+    std::vector<chainerx::Array> params_;
+};
+
+chainerx::Array SoftmaxCrossEntropy(const chainerx::Array& y, const chainerx::Array& t, bool normalize = true) {
+    chainerx::Array score = chainerx::LogSoftmax(y, 1);
+    chainerx::Array mask =
+            (t.At({chainerx::Slice{}, chainerx::NewAxis{}}) == chainerx::Arange(score.shape()[1], t.dtype())).AsType(score.dtype());
+    if (normalize) {
+        return -(score * mask).Sum() / y.shape()[0];
+    }
+    return -(score * mask).Sum();
+}
+
+void RunWithDefaultDevice(int64_t epochs, int64_t batch_size, int64_t n_hidden, int64_t n_layers, float lr, const std::string& mnist_root) {
+    // Read the MNIST dataset.
+    chainerx::Array train_x = ReadMnistImages(mnist_root + "train-images-idx3-ubyte");
+    chainerx::Array train_t = ReadMnistLabels(mnist_root + "train-labels-idx1-ubyte");
+    chainerx::Array test_x = ReadMnistImages(mnist_root + "t10k-images-idx3-ubyte");
+    chainerx::Array test_t = ReadMnistLabels(mnist_root + "t10k-labels-idx1-ubyte");
+
+    train_x = train_x.AsType(chainerx::Dtype::kFloat32) / 255.f;
+    train_t = train_t.AsType(chainerx::Dtype::kInt32);
+    test_x = test_x.AsType(chainerx::Dtype::kFloat32) / 255.f;
+    test_t = test_t.AsType(chainerx::Dtype::kInt32);
+
+    int64_t n_train = train_x.shape().front();
+    int64_t n_test = test_x.shape().front();
+    chainerx::Array train_indices = chainerx::Arange(n_train, chainerx::Dtype::kInt64);
+
+    // Initialize the model with random parameters.
+    Model model{train_x.shape()[1], n_hidden, 10, n_layers};
+    std::random_device rd{};
+    std::mt19937 gen{rd()};
+    std::normal_distribution<float> dist{0.f, 0.05f};
+    model.Initialize(gen, dist);
+
+    auto start = std::chrono::high_resolution_clock::now();
+
+    for (int64_t epoch = 0; epoch < epochs; ++epoch) {
+        // The underlying data of the indices' Array is contiguous so we can use std::shuffle to randomize the order.
+        assert(train_indices.IsContiguous());
+        std::shuffle(
+                reinterpret_cast<int64_t*>(train_indices.raw_data()), reinterpret_cast<int64_t*>(train_indices.raw_data()) + n_train, gen);
+        for (int64_t i = 0; i < n_train; i += batch_size) {
+            chainerx::Array indices = train_indices.At({chainerx::Slice{i, i + batch_size}});
+            chainerx::Array x = train_x.Take(indices, 0);
+            chainerx::Array t = train_t.Take(indices, 0);
+
+            chainerx::Backward(SoftmaxCrossEntropy(model(x), t));
+
+            // Vanilla SGD.
+            for (const chainerx::Array& p : model.params()) {
+                chainerx::Array p_as_grad_stopped = p.AsGradStopped();
```

Comment:
How about aligning variable names to the Python example? `p_as_grad_stopped` sounds too verbose as an example.

Reply:
Fixed.